### DISCUSSION TITLE
Review/Merge of my PRs

### DISCUSSION THREAD

**Discussion by batzen** [dotnet/wpf#6542] [General]
Hi there.

I start to get frustrated as not a single of my PRs (dating back to 07.2019) has been fully reviewed or been merged.
I would appreciate some feedback from the WPF team.

To get a list of my PRs: https://github.com/dotnet/wpf/pulls/batzen

Despite pinging @dotnet/wpf-developers some weeks ago, i did not receive any feedback.
Because of that i ping @SamBent @dipeshmsft directly here now.

Would it at least be possible to get feedback on why my PRs don't get reviewed, while other community contributions get reviewed and some even get merged?

**Comment by gurpreet-wpf** [comment]
Hi @batzen,

We are working to close the PRs from all the contributors in the community. Performance PRs generally have a longer test/verification cycle and they are currently in lower priority as opposed to bug fix PRs. It is coincidental that none of your PRs have been merged recently. However, we would review and update the PRs submitted by you. Thanks for your patience!

**Reply by batzen** [reply]
@gurpreet-wpf Is there anything the community can do to help the WPF team?

**Reply by gurpreet-wpf** [reply]
@batzen We are going to start community-run projects and an announcement has been made regarding the same under Discussions section. Looking forward to suggestions from the wider community.

**Comment by wstaelens** [comment]
@batzen the WPF team doesn't look at issues nor pull requests. 
I guess they are only 2 or 3 people working (or sleeping) here in the @dotnet/wpf-developers team . @danmoseley really has some work to manage his teams. Some bugs are already out there for years.
I've only seen them putting "Future" labels on items or "closing" them because they could not reproduce or didn't have more info. That is the easy and lazy way of dealing with stuff.


We also have some big frustrations regarding open issues on WPF/XPS. 

@batzen  even now you are being pleased with a message "thanks for your patience", they just don't seem to care about WPF/XPS.

**Reply by batzen** [reply]
@wstaelens While i get your frustration, being harsh/bitter and adding accusations only adds harm for no professional reason.

**Reply by wstaelens** [reply]
@batzen sometimes after 4? 5? 6? years of nothing being fixed and still encountering the same issues in a new codebase (.NET Framework vs. .NET Core) you often get frustrated and say things hoping it would change "something". But I agree, sorry for that.

**Comment by bgrainger** [comment]
I share this frustration, having a couple of PRs of my own that have not been merged nor received any feedback.

There are now two community projects (https://github.com/dotnet-campus/dotnetCampus.CustomWpf/blob/dev/README.md, https://github.com/Faithlife/wpf/blob/stable/README.md) that are shipping forks of WPF that include some fixes the WPF team has not yet merged into `main`.

The maintainers of each of these (I am one of them) have gone through a number of open PRs and quickly determined that the proposed changes provide an improvement with a low risk of regression. The list of merged PRs in each is [here](https://github.com/dotnet-campus/dotnetCampus.CustomWpf/issues/10) and [here](https://github.com/Faithlife/wpf/blob/stable/ReleaseNotes.md). I really don't understand why the WPF team can't do the same, instead of [asking the community to upvote PRs they've already made](https://github.com/dotnet/wpf/discussions/6556). I believe there are a number of people who would eagerly test `7.0.0-preview` releases of WPF with their applications to help ensure no critical regressions have been introduced from merging these PRs.

I'd much rather put effort into improving WPF itself than packaging and shipping my own fork, but if the latter is the only way I can actually deliver fixes to customers of my application, it's unfortunately what I have to do right now.

**Reply by wstaelens** [reply]
🔔 🗣️  @SamBent @predavid @vishalmsft @danmoseley @shanselman @KirillOsenkov @gurpreet-wpf @ryalanms ?

**Comment by Wraith2** [comment]
This issue closed 2 years ago covers the same ground. https://github.com/dotnet/wpf/issues/2884

My conclusion was that the source is open source but the development is not. There was a year where there was no team and even 2 years later when there has been a team there was no continuity of developers from the original to the new. So there's a new team with a large complex codebase they don't know well and no-one to ask about the history. In that position it's hard for anyone to succeed.

Forking may be the best idea to keep the project alive. Investment is not going to come from Microsoft, this has been demonstrated. Management and merge authority cannot be granted to anyone outside Microsoft.

**Reply by wstaelens** [reply]
Sounds like thé Windows Phone story... MS whining about the lack of apps and low adoption rate, but not doing their best to port popular apps and not providing enough examples or listening to the community. When it is too late and the boat sunk, they reacted...too late. (Even the WP program manager Joe Belf., didn't believe and really used/promoted the Windows Phone, he was rather into iPhone/Android. The community didn't see a leader using it's own products and the result....is history)
The WPF community is shouting now again... Hope somebody listens.

**Reply by janseris** [reply]
Sounds like Xamarin story as well. I don't understand why Microsoft doesn't put resources into bugfixes of these platforms. As if Microsoft didn't have money for that

**Comment by Symbai** [comment]
It has absolutely no sense to talk to the WPF repo owners because they just tell you "We will get to that soon, we just need to do that first". We are hearing this statements since WPF has been open sourced. Also the 72 hours per week to address issues or reviewing PRs is a lie. I've watched the repo from Monday to Thursday just to see not a single new issue or PR has been addressed, not even labeled. This is not harsh and not bitter, its just naming the facts. And we have to be very clear about the issue to demonstrate how big it is.

Anyway you need to find the man or woman who was responsible for assigning the WPF repo to this team and let him know that it is not going to work. I don't know the name of this person but posting it here is a very less chance to change something. You need to raise your words on for example .NET announcement blogs. I know that at least some people with bigger cross repo responsibility are reading that for sure.

**Reply by wstaelens** [reply]
@Symbai i don't expect anyone of them to actually make some moves at the WPF team (i hope so but ...)
Maybe you can try with recent authors of the devblogs: @tannergooding @poppastring @luisquintanilla @AngelosP @JeremyLikness @danroth27 @TaraOverfield @mjrousos ?

**Reply by Symbai** [reply]
No, most of them have nothing to do with it. I had a chance talking shortly with Rich Lander about it. He said he also wishes that there would be more activity here and when I mentioned the great afford of some people trying to add NativeAOT and showed him the PR links to it he said they were good examples but in the end he refused to discuss "internal decisions on resources" any further, putting that in quotes not because it was his exactly words but to summarize it.

We need to find someone being responsible or someone knowing someone else being responsible but we shouldn't go around and talk directly to random persons. Because THIS would be bitter. Imagine you work for a company and a random person comes to you and trash talk about other employers work. Not a nice move at all.

I just did that post above only because I'm seeing this problem since the open source, trying to mention it everywhere and made the experience that its not helpful to mention it here at all. For the reasons above. My frustration already has turned into resignation. I wish @batzen good luck but I'm very sure his PRs remain at their state for the next months for sure. And the recent announcement to work on upvoted PRs but ONLY on bug fixes, is another good indicator that I might be very right. And I truly wish people would not post here but also on other places. The more people do, the more helpful it is. But if you see only one comment out of 50 asking about WPF's future and its current problems, that's making it look like very low priority 🤷‍♂️

**Reply by Wraith2** [reply]
> in the end he refused to discuss "internal decisions on resources" any further, putting that in quotes not because it was his exactly words but to summarize it.


This is reasonable if it was not his decision. 

So, how do we find whose decision it was and get our feedback to them? We don't really know who owns what because we're outside Microsoft and whoever has made a decision about the relative worth of wpf for resourcing is unlikely to spend their time on github projects so we're unlikely to catch them spraying @'s about.

@richlander , @danmoseley as visible managers could you help us get this discussion in front of the people who can help us understand the current state of the project please? The community is trying to engage and push the project forward and is unable to do so.

**Reply by tannergooding** [reply]
I'd recommend not pinging the authors of recent blog posts or other publicly visible devs in general, many of us work in areas completely disconnected from WPF and may not know where to start here ;)

Of all the people pinged, Rich and Dan are probably the closest to someone in a position to help here but it doesn't fall under their purview/ownership either.

**Reply by richlander** [reply]
I haven't talked to the WPF Team recently. I'll setup some time with them to better understand where they are at. I understand why folks are frustrated and sympathize.

**Comment by gurpreet-wpf** [comment]
Thank you all for being contributors to WPF and helping make the framework better every day. We have reviewed the comments and acknowledge the frustration on not getting timely feedback on your contributions and failing to meet your expectations so far. We are taking steps to address this problem.

Some of the PR(s) touch a lot of areas (sometimes to the very fundamentals), whereas some help performance. These PR merges require extra diligence from our side to ensure product integrity on older platforms. With manual testing, diagnosis of the root cause in the case of a test failure becomes challenging. Hence you see a relatively long lead time for feedback.

Here is a little more context - Unfortunately, WPF product suite does not have unit tests in place. We have 30K+ integration tests that take more than 18+ hours to complete. These integration tests are manually run on various OS versions to ensure that the framework is fully backward compatible with all older supported platforms. For every change that does make it to the main branch, we run these tests to ensure integrity of the product. We acknowledge there are opportunities for improvement here and have increased investment in this area.

Hence, the next logical step for us is to have test automation in place and have a subset of these tests run on every incoming PR, thereby improving the turn-around times on PRs. We are focused on this, to make sure our test infrastructure is modernized and tests automated in a way that community receives feedback in a timely manner.

We have always chosen to accept as many community PRs as we can and will aim to reduce the time for feedback via Test Automation. We humbly request our community contributors to not be disheartened and not give up on us just yet. We'll be publishing our goals for this year in a separate post by the end of this month.

**Reply by richlander** [reply]
I have two questions that might help the community.

1. If you decide to start the process with an issue, do you mark the PR in some way to say that the process has started to consider accepting the PR?
2. If the test pass fails, do you share that result, including the errors with the PR author?

Here's an example of a project I work on. We write "triage" messages to explain our thinking to issue/PR authors. https://github.com/dotnet/dotnet-docker/issues/3844#issuecomment-1156785469

**Reply by gurpreet-wpf** [reply]
@richlander 
1. As of now, we don't mark the PRs when we are reviewing them. However, an extra label gets added once the PR is selected for Test Pass. We are looking to add more labels to improve the visibility.
2. Currently we don't, the reason being that our test infrastructure is internal. So, any change that need to be made to the test infrastructure is done by us. We are planning to improve test results visibility in upcoming releases.

**Reply by richlander** [reply]
We have an internal meeting coming up to talk about this, but I can share my initial feedback now. Please don't use labels as the primary way to to talk to contributors. It's objectively an anti-pattern. People CRAVE direct engagement, as *clearly* demonstrated on this thread. I get it. I'm like that, too. Zero people are interested in interacting with a state transition machine.

The thing people want is something like this:

> Thanks for your contribution. We're going to run it through a test pass now. You can expect another response from us within a week.

or

> Thanks for the contribution. I like the idea, but we cannot take this change as-is. The challenge I see is blah.

or 

> Thanks for the contribution. I see where you are headed. I've got specific feedback that I'll you directly on your changes.

**Reply by kant2002** [reply]
That's kind of conversations which is really turn me contributing to NativeAOT. Also I would like to find a ways to reduce time between file PR and time when it lands. 2 weeks for trivial contributions is not motivate very much. Even if I understand not willingness to risk, I will to reduce risk by writing tests, if that helps land my PR faster. Tests which would be public and every contributor can run.

**Reply by bgrainger** [reply]
> These integration tests are manually run on various OS versions to ensure that the framework is fully backward compatible with all older supported platforms.

Is some of that work now unnecessary, if WPF is targeting .NET 7.0 and .NET 7.0 is dropping support for older OSes? https://github.com/dotnet/announcements/issues/226

(It still has to be tested back to Windows 10 1607, though.)

**Reply by kant2002** [reply]
There small windows where .NET 7 will support Windows 7 and Windows 8.1, from Nov 2022 till Jan 2023. At least that's what WinForms would [do](https://github.com/dotnet/winforms/pull/7087#discussion_r873284879). But for .NET 8 we can drop everything not Win 10 from my understanding.

**Reply by danmoseley** [reply]
The Windows 7/8.1 support timeframe in https://github.com/dotnet/announcements/issues/226 applies to all of the .NET product (inevitably so as it all runs on the runtime/core libraries)

**Reply by ThomasGoulet73** [reply]
> Please don't use labels as the primary way to to talk to contributors.

I would also like to add that adding labels does not trigger Github notifications (I don't think you can even configure Github to receive notifications on label change). Adding a comment when PRs are under review, queued for test pass, etc. would send notifications which is useful for those that don't want to check their PRs for updates.

**Reply by pchaurasia14** [reply]
@ThomasGoulet73 - Absolutely. Labels only help us group items and plan internally. However, we will definitely be adding more comments to the PR/Issues we'd working on.

**Comment by Wraith2** [comment]
> We have reviewed the comments and acknowledge the frustration on not getting timely feedback on your contributions and failing to meet your expectations so far. We are taking steps to address this problem.

That's good but we have been given promises before and no results were seen. How will we be able to measure the progress of these steps to address the problem, what can we expect to see change and in what timescales will we be able to judge success? 

> Hence, the next logical step for us is to have test automation in place and have a subset of these tests run on every incoming PR, thereby improving the turn-around times on PRs.

This seems like a very clear issue to open up to community contributions. While no-one enjoys paying off accrued technical debt it is something that a larger force of dedicated volunteers can work on in parallel. If a test harness is provided a rule that no PR can pass without covering tests could be made and in that way coverage increases with community and dedicated involvement.
Has the automation of integration testing been a target of sustained development by the team? If so what progress has been made (because there is nothing in the repository as far as I know)? If not then can you explain why not?

The owners from the Microsoft dedicated team have the sole capability to bottleneck progress. The community can only help with their assistance. This has both political-technical (ownership and use of wpf) and practical (github capabilities) roots and will not change. 

There are are many people who are willing and able to help improve the project and I think that you need to give unblocking their ability to contribute a higher priority. These people are willing to do difficult, complicated, time consuming, work for free, that is a resource pool worth treasuring not appearing to ignore. 

> We humbly request our community contributors to not be disheartened and not give up on us just yet. 

Communication with the community is also an area that needs work. With regular communication the relative importance of each interaction decreases and the perceived availability of the owners increases improving feeling of inclusion. Even if it's just a note that a test would be a good idea or a trivial style comment on the code the interaction itself is important to show that the github repo is not a low priority task to be done when there is free time but is instead the location where all contribution effort is centred. 

Some other repositories in the dotnet organisation do this very well and it might be worth taking some time to see how they work with their communities to create the feeling of joint development on a project instead of leaving their community feeling like a neglected afterthought.

> We'll be publishing our goals for this year in a separate post by the end of this month.

I'll look forward to seeing the goals and hope that there are tangible ways to measure the progress towards them.

**Reply by pchaurasia14** [reply]
> How will we be able to measure the progress of these steps to address the problem, what can we expect to see change and in what timescales will we be able to judge success?

We're planning to leverage Github projects to enable granular progress monitoring on this front. For the timescales, we'll be publishing it along with the goals in our upcoming post.


> This seems like a very clear issue to open up to community contributions.

Exactly. Hence, we are working at full speed to get the basic test framework out in the open as soon as possible. In fact, [wpf-test](https://www.github.com/dotnet/wpf-test) already contains a very small subset of basic tests. We'll be enriching that repository with more tests in upcoming sprints.


> Has the automation of integration testing been a target of sustained development by the team? If so what progress has been made (because there is nothing in the repository as far as I know)? If not then can you explain why not?

Yes. Whilst working to open source the tests, it is equally important for us to be able to run these tests in automated fashion. Since the automation pipeline requires use of artifacts that are not open-sourced yet, we haven't released anything on Github yet. Pipelines will be exposed to community when the required infrastructure is made available. 


> These people are willing to do difficult, complicated, time consuming, work for free, that is a resource pool worth treasuring not appearing to ignore.

Totally. Together we can achieve unprecedented things. 😎

> Communication with the community is also an area that needs work. 

Feedback noted. We will definitely work on improving our feedback both in terms of frequency and in terms of intensity so that community does not feel left out. 

> I'll look forward to seeing the goals and hope that there are tangible ways to measure the progress towards them.

Stay tuned and thank you so much for all the feedback.! 😊

**Reply by richlander** [reply]
> the repo is location where all contribution effort is centred

I love this. Definitely agree.

**Reply by pchaurasia14** [reply]
As promised, we've published our half yearly goals [here](https://github.com/dotnet/wpf/discussions/6744). 

We'd love to have your feedback, comments and suggestions. 😊

**Comment by kant2002** [comment]
What about other important PR's which is not simple bug fixes, for example like these one https://github.com/dotnet/wpf/pull/6171
That's kind of work or maybe untangling circular dependencies are important to make WPF more easy to work with and modernize.
This is hard PR to communicate, but they are still should be looking onto. If internal test suite is enough to be sure that nothing is breaking, maybe they can be considered for testing as well?

**Reply by pchaurasia14** [reply]
Absolutely @kant2002. 

We cannot wait to have this PR reviewed and merged. In fact, it is changes like these that we want to absorb as soon as possible.  Like you mentioned such changes are not easy to merge without thorough testing. Presumably, if such changes breaks any of our tests, then finding out where to fix it (either in the tests itself or the PR) would take relatively longer if it is done by us alone. 

Hence, we are concentrating on test migration and test infrastructure setup, so that such things do not get blocked on us. 
Rest assured, we are definitely going to look into this as soon as we can.

Oh and BTW, if it has not already been mentioned, A BIG THANK YOU for all your contributions. 😎

**Comment by Symbai** [comment]
So according to the published goals, reviewing and merging of community PRs does not happen before year 2023. That makes .NET 7 out of scope. And some PR's will be more than 5 (FIVE) years old then. If it really takes FIVE years (= 1,300 working days = 10,400 hours) only to setup testing, then I really, really don't know what to say. At least I know that I cannot say "Hurraay" to that. But if it makes the contributors happy... it's okay then I guess?!

**Reply by wstaelens** [reply]
@Symbai I had the same mixed feelings, again promises and "Future"-tags.

**Reply by Wraith2** [reply]
I was taking some time to try and frame my thoughts politely. I am dubious plans because of the lack of defined measurable goals, with 1300 tests are we just supposed to count the ones added to the tests repo and celebrate when it hits 13000? I'm Disappointed that the goals are so far away and largely obvious things that we're already talked about here.

When I thought it through I realised that no matter what the plan or response is I'm unlikely to be satisfied. As we've pointed out in this discussion there have been plans and platitudes before but nothing has happened. The only thing that will satisfy me that anything has changed is concrete results, actually delivering on _anything_ promised. It is too late for trust because I have been disappointed too many times. Based solely on actions we see this is an inactive project outsourced to an offshore maintenance team.

Deliver on promises or stop pretending this is an active project. More promises are not sufficient.

Oh, and stop thanking people for their contributions when they're not being accepted. It's obnoxious. I don't want "managing" I want an honest dialog.

**Reply by richlander** [reply]
There may be some misunderstanding on "FY23". Fiscal year 2023 (for Microsoft) started on July 1st 2022. Certainly, the poster should have shared that context. In general, we try to avoid Microsoft specific jargon on GitHub. Using fiscal years isn't useful for community engagement.

**Reply by gurpreet-wpf** [reply]
Sorry for the confusion. We have clarified the dates in the original post. Thanks.

**Reply by Wraith2** [reply]
We're most of the way through FY23 now and while it looks like some things have moved there are still a lot of community contributions open. Looking through the 4 pages of open pull requests 90% of them are community contributions sitting waiting for attention. So it doesn't look like a lot as moved in terms of community engagement.

**Reply by pchaurasia14** [reply]
@Wraith2 - Yes, you're right. We have a lot of ground to cover before we can review/merge all the community PRs. Unfortunately, we are also slowed down by the quality of tests that can safely access the impact of the changes that go in the product. (I wish I had better news).

That being said, we continue to review and [merge](https://github.com/dotnet/wpf/discussions/7698) community PRs at a steady cadence.

**Reply by kant2002** [reply]
How many tests are left to be ported? In percentages and approximate. Right now community held as hostages and should wait while something and out of our control has to be done. We still cannot guess when this process finished. @pchaurasia14 can you get some insights on this?

**Reply by pchaurasia14** [reply]
@kant2002 -  We have roughly migrated around ~6K tests (including ones that are about to be released by this week) of 20K+ tests (30%).  We are working to ensure that bulk of the tests are made available as soon as possible.

**Comment by Symbai** [comment]
I DO see activity from the WPF team here in this repo. They are trying to fix (less important / most of the time not reported here on github, or self introduced) issues, but after 6 months I still see @batzen PRs not reviewed or merged and no progress on other (important) PRs as well, for example #6171 , so the term of "FY23" was definitely not the problem. But after 5 years, what matters waiting another year, right!? See you in 2023, have a great holiday.

**Reply by wstaelens** [reply]
@Symbai i opened a ticket on runtime just so somebody would actually read and act: https://github.com/dotnet/runtime/issues/78629

Many frustrations from our side here.

@Symbai : https://devblogs.microsoft.com/dotnet/wpf-on-dotnet-7/comment-page-2/#comment-17219  leave some comments here.

**Reply by Symbai** [reply]
I understand your frustration, trust me. But please don't comment in other PRs just to point to an issue you'd like to have fixed.  It's not fair to others. If we all would doing this things would get out of control. And you wouldn't like of others doing the same in your issues / PRs.

**Reply by wstaelens** [reply]
Yes correct but it seems to be the only way to drive the team some attention. and yes i would love to see your and other people's issues being fixed also!! "things would get out of control", i don't think there is any good control right now, as bugs are open for years. 

Hope they will understand the need for a bigger WPF team.

**Reply by Wraith2** [reply]
A larger team seems unlikely to happen. /runtime as well as other active and well engaged repositories are part of devdiv but it seems like other divisions aren't as well adjusted or willing to work in the same way with the community. What you're seeing is open source but closed everything else. The bugs that are being fixed are the ones that are reported internally and the issues here are treated as a secondary source. 

It is unfortunate that the oss way to "fix" this by forking isn't really applicable for first party libraries in the .net ecosystem because the level of trust and support required by commercial consumers means that uptake of the fork would be low.

**Reply by Wraith2** [reply]
I was bored so I went looking through the commit history for the repo. I had to go back 17 pages before I found a genuine community PR, there were others but they were all by community triagers and many look like internal issues from MS for things like VS. In general the most activity in the repo is by auto merge bots for dependencies. There seems to be little to no progress on issues.

**Reply by wstaelens** [reply]
just as I thought... excuses, promises but in the end it gets a label "future" to keep you busy.
